### DISCUSSION TITLE
3d Mouse Support

### DISCUSSION THREAD

**Discussion by Mcmartelle** [hannobraun/fornjot#1666] [Ideas]
Awesome project! Is 3d mouse support on the radar?

My 3dconnexion SpaceMouse can zoom in and out by pretending to be a scroll wheel, but the other axis don't seem to have an interface, besides regular mouse clicks and drags.

I saw there's a crate called spacenav that claims to talk to an open source daemon which can talk to the 3dconnexion software, but it's 6 years old. I'd be willing to see if this approach works.

What do you think?

**Comment by hannobraun** [comment]
Hey @Mcmartelle, thank you for your interest!


> Awesome project!

Thank you!


> Is 3d mouse support on the radar?

Not before you wrote this :smile: 

I know it's a thing, but I've never used one nor thought much about it. Given my workload, it's extremely unlikely that  this is something I'll look into anytime soon, so any work in this area will have to be driven by contributors.


> I saw there's a crate called spacenav that claims to talk to an open source daemon which can talk to the 3dconnexion software, but it's 6 years old. I'd be willing to see if this approach works.
> 
> What do you think?

In principle I would welcome this feature, but there are a few points to keep in mind:

1. I don't have a SpaceMouse, and this is unlikely to change anytime soon. So any continued maintenance would have to be performed by contributors too.
2. If SpaceMouse support breaks or causes issues with other code, and this can't be resolved in a timely manner, I might decide to (temporarily) remove the feature.
3. Typically, I'm happy to merge code that isn't ideal, if it's something that I know I'll end up working on eventually anyway. Since this is code I'm unlikely to touch and might want to remove, there will be a somewhat higher bar when it comes to modularity. Basically, this needs to be its own module, easily separated from other input code.

If this is fine by you, then it's fine by me! I'd be happy to merge this feature. I just want to manage expectations in regard to the need of ongoing maintenance.

**Reply by Mcmartelle** [reply]
Thanks for replying.

Making a 3d mouse module separate from the input code makes a lot of sense.

I know a version of the official 3dconnexion software allowed custom mapping of each axis to a keyboard or mouse output, on an app by app basis. If that old spacenav crate doesn't work, we may only need to give the 3d mouse another input to emulate, like gamepad joysticks.

Anyways, I'll give it a shot.

**Reply by hannobraun** [reply]
Sounds good! Looking forward to what you come up with!

**Comment by xpe** [comment]
@hannobraun Given the [change of focus](https://www.fornjot.app/blog/a-new-direction/) it seems like we might want to <s>close this issue for housekeeping purposes?</s> Ah, here we are in Github Discussions, not Issues :)... Do you want to mark this discussion as "out of scope" or similar?

**Comment by hannobraun** [comment]
Good point, @xpe, thank you! I usually don't think much about closing discussions, but you're right, no reason to keep this open.
